clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -752,6 +758,106 @@ private StoragePartitionConfig getStoragePartitionConfig(
     return storagePartitionConfig;
   }
 
+  /**
+   * HeartBeat SOS messages carry the leader completion state in the header. This function extracts the leader completion
+   * state from that header and updates the {@param partitionConsumptionState} accordingly. <p>
+   * If there is no leader completion state header, reset the leader completion state to
+   * {@link LeaderCompleteState#LEADER_COMPLETE_STATE_UNKNOWN} as the leader don't know about this header
+   * (using old version of the code) or the leader may have rolled back to a version that doesn't support this header or
+   * this topic partition gets a new leader which doesn't support this header yet.
+   */
+  protected void getAndUpdateLeaderCompletedState(
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+      PartitionConsumptionState partitionConsumptionState) {
+    if (isDaVinciClient || partitionConsumptionState.getLeaderFollowerState().equals(STANDBY)) {
+      ControlMessage controlMessage = (ControlMessage) consumerRecord.getValue().payloadUnion;
+      ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+      if (controlMessageType == ControlMessageType.START_OF_SEGMENT
+          && Arrays.equals(consumerRecord.getKey().getKey(), KafkaKey.HEART_BEAT.getKey())) {
+        boolean isLeaderCompleteHeaderFound = false;
+        LeaderCompleteState oldState = partitionConsumptionState.getLeaderCompleteState();
+        LeaderCompleteState newState = oldState;
+        PubSubMessageHeaders pubSubMessageHeaders = consumerRecord.getPubSubMessageHeaders();
+        for (PubSubMessageHeader header: pubSubMessageHeaders.toList()) {
+          if (header.key().equals(VENICE_LEADER_COMPLETION_STATE_HEADER)) {
+            newState = LeaderCompleteState.valueOf(header.value()[0]);
+            partitionConsumptionState
+                .setLastLeaderCompleteStateUpdateInMs(consumerRecord.getValue().producerMetadata.messageTimestamp);
+            isLeaderCompleteHeaderFound = true;
+            break; // only interested in this header here
+          }
+        }
+        if (!isLeaderCompleteHeaderFound) {
+          // reset LeaderCompleteState: If a leader originally sent this header but later is rolled back to a version
+          // that doesn't support this header or this TP gets a new leader which doesn't support this header yet.
+          newState = LEADER_COMPLETE_STATE_UNKNOWN;
+        }
```

Comment:
So for AA store, when we do a colo by colo rollout, servers with old codes would not produce this header, and thus followers with new codes would be back and forth on leader complete state. In theory, it's okay, right?

Reply:
This will be needed only until all the servers are running the latest version. During transition, my understanding is that it's okay if old or the new way is used to decide the follower completion status.